### ISSUE TITLE
Use as SWAP

### ISSUE THREAD

**Issue by snshn** [Overv/vramfs#3]
Was wondering if it could be possible to host a swap partition within vramfs or somehow patch vramfs to make it work as a swap partition?

My drive is encrypted, therefore I don't use SWAP partitions... but if this thing could give me 3GB or so of a swap-like fs, we could be onto something...

Do you think it could work without fuse, natively?

Oh, and great idea behind vramfs, really neat!

**Comment by tarcey** [comment]
It already is possible to use a vramfs for swap! 
I was reading the Arch Wiki about swap which states that there is a dangerous method (can corrupt fs, not really a bin deal with vramfs) to get swap files working in BTRFS. So i found this:
http://www.spinics.net/lists/linux-btrfs/msg28533.html (A thread that shortly explains how to create a swap file in BTFS, the same method works for vramfs)
I tried it and i can confirm it works. You will need to have a Kernel with loopdevice support, you usually should already have it, if not, build a Kernel with loopdevice support ( if i remember correctly the option resides somewhere under Device Drivers -> Block Devices, but i might be wrong)

**Comment by bisqwit** [comment]
While it is technically possible to create a file on VRAMFS and use it as a swap, this is risky: What happens if VRAMFS itself, or one of the GPU libraries, gets swapped? This can happen in a low-mem situation, i.e exactly in a situation that swap is designed to help. The kernel cannot possibly know that restoring data from the swap depends on the data that is… swapped in the swap.
This is not an issue for kernel-space filesystem/storage drivers, but it is a conundrum for user-space stuff.

**Comment by j123b567** [comment]
For kernel-space driver, it would be nice to use directly TTM/GEM to allocate video ram buffers.

**Comment by bisqwit** [comment]
What are TTM/GEM?

Note that the slram/phram/mtdblock thing can only access at most like 256 MB of the memory, the size of the memory window (I guess) of the PCI device.

**Comment by j123b567** [comment]
I don't know much, but they are some interfaces to acces GPU memory inside kernel. So it can see all the GPU memory, not only some mapped part directly accesible. https://www.kernel.org/doc/html/latest/gpu/drm-mm.html

My situation, NVidia dedicated GPU with 4GB RAM and nouveau driver without OpenCL support. This memory is not mapped to memory space so I can't use them using slram/phram.

**Comment by dhalsimax** [comment]
> It's possible to implement a block device with OpenCL backing it. It could probably be developed pretty quickly with something like [BUSE](https://github.com/acozzette/BUSE).

The easy way to accomplish is to use vmrafs as is, make a file on vramfs disk then use a loop device on that file, format the loop device with mkswap and then swapon. With this method everything seems to work as I tried. Anyway the big issue using FUSE or BUSE is that both runs in user space and user space is swappable. I have not tried it, but suppose the memory  of the vramfs process get swapped itself by the kernel, how would the kernel be able to recover by a page fault as it needs to reload in the first place ? I am curious what will happen then?

**Comment by LHLaurini** [comment]
> What happens if VRAMFS itself, or one of the GPU libraries, gets swapped?

Couldn't [`mlockall`](https://man7.org/linux/man-pages/man2/mlock.2.html) be used to prevent vramfs from getting swapped?

**Comment by montvid** [comment]
Wonderful idea! I am runnnig an old headless server with a 1 gb ddr3 amd card opencl 1.1. I can use all the video ram as i use just ssh. Unfortunately vramfs does not let me create a swap file based swap I get "swapon: /mnt/vram/swapfile: swapon failed: Invalid argument". Can it be fixed? I see opencl 1.2 is merged into mesa 20.3 so good times ahead for this project.

**Comment by wonghang** [comment]
It doesn't work for me. Even I tried to mlockall() page for the userspace program. I think the nvidia driver allocated some memory that would be swapped. At some point, the computer will get into deadlock when memory is low.

I also tried the BUSE / nbd approach. It doesn't work for me as well.

I think we need to get into the nvidia driver, carefully develop a block device kernel driver and call these undocumented API:

```
cat /proc/kallsyms |grep rm_gpu_ops | sort -k 3
0000000000000000 t rm_gpu_ops_address_space_create	[nvidia]
0000000000000000 t rm_gpu_ops_address_space_destroy	[nvidia]
0000000000000000 t rm_gpu_ops_bind_channel_resources	[nvidia]
0000000000000000 t rm_gpu_ops_channel_allocate	[nvidia]
0000000000000000 t rm_gpu_ops_channel_destroy	[nvidia]
0000000000000000 t rm_gpu_ops_create_session	[nvidia]
0000000000000000 t rm_gpu_ops_destroy_access_cntr_info	[nvidia]
0000000000000000 t rm_gpu_ops_destroy_fault_info	[nvidia]
0000000000000000 t rm_gpu_ops_destroy_session	[nvidia]
0000000000000000 t rm_gpu_ops_device_create	[nvidia]
0000000000000000 t rm_gpu_ops_device_destroy	[nvidia]
0000000000000000 t rm_gpu_ops_disable_access_cntr	[nvidia]
0000000000000000 t rm_gpu_ops_dup_address_space	[nvidia]
0000000000000000 t rm_gpu_ops_dup_allocation	[nvidia]
0000000000000000 t rm_gpu_ops_dup_memory	[nvidia]
0000000000000000 t rm_gpu_ops_enable_access_cntr	[nvidia]
0000000000000000 t rm_gpu_ops_free_duped_handle	[nvidia]
0000000000000000 t rm_gpu_ops_get_channel_resource_ptes	[nvidia]
0000000000000000 t rm_gpu_ops_get_ecc_info	[nvidia]
0000000000000000 t rm_gpu_ops_get_external_alloc_ptes	[nvidia]
0000000000000000 t rm_gpu_ops_get_fb_info	[nvidia]
0000000000000000 t rm_gpu_ops_get_gpu_info	[nvidia]
0000000000000000 t rm_gpu_ops_get_non_replayable_faults	[nvidia]
0000000000000000 t rm_gpu_ops_get_p2p_caps	[nvidia]
0000000000000000 t rm_gpu_ops_get_pma_object	[nvidia]
0000000000000000 t rm_gpu_ops_has_pending_non_replayable_faults	[nvidia]
0000000000000000 t rm_gpu_ops_init_access_cntr_info	[nvidia]
0000000000000000 t rm_gpu_ops_init_fault_info	[nvidia]
0000000000000000 t rm_gpu_ops_memory_alloc_fb	[nvidia]
0000000000000000 t rm_gpu_ops_memory_alloc_sys	[nvidia]
0000000000000000 t rm_gpu_ops_memory_cpu_map	[nvidia]
0000000000000000 t rm_gpu_ops_memory_cpu_ummap	[nvidia]
0000000000000000 t rm_gpu_ops_memory_free	[nvidia]
0000000000000000 t rm_gpu_ops_own_page_fault_intr	[nvidia]
0000000000000000 t rm_gpu_ops_p2p_object_create	[nvidia]
0000000000000000 t rm_gpu_ops_p2p_object_destroy	[nvidia]
0000000000000000 t rm_gpu_ops_pma_alloc_pages	[nvidia]
0000000000000000 t rm_gpu_ops_pma_free_pages	[nvidia]
0000000000000000 t rm_gpu_ops_pma_pin_pages	[nvidia]
0000000000000000 t rm_gpu_ops_pma_register_callbacks	[nvidia]
0000000000000000 t rm_gpu_ops_pma_unpin_pages	[nvidia]
0000000000000000 t rm_gpu_ops_pma_unregister_callbacks	[nvidia]
0000000000000000 t rm_gpu_ops_query_caps	[nvidia]
0000000000000000 t rm_gpu_ops_query_ces_caps	[nvidia]
0000000000000000 t rm_gpu_ops_release_channel	[nvidia]
0000000000000000 t rm_gpu_ops_release_channel_resources	[nvidia]
0000000000000000 t rm_gpu_ops_report_non_replayable_fault	[nvidia]
0000000000000000 t rm_gpu_ops_retain_channel	[nvidia]
0000000000000000 t rm_gpu_ops_retain_channel_resources	[nvidia]
0000000000000000 t rm_gpu_ops_service_device_interrupts_rm	[nvidia]
0000000000000000 t rm_gpu_ops_set_page_directory	[nvidia]
0000000000000000 t rm_gpu_ops_stop_channel	[nvidia]
0000000000000000 t rm_gpu_ops_unset_page_directory	[nvidia]
```

to create a GPU session and allocate GPU memory in order to make a GPU swap truly possible.

**Comment by tugrul512bit** [comment]
I wrote a header-only multi-gpu C++ virtual object array implementation. Would it help for you to keep data safely? It doesn't use mlock/mlockall but Opencl's own locking mechanism (alloc_host_ptr flags and map/unmap).

https://github.com/tugrul512bit/VirtualMultiArray

**Comment by azureblue** [comment]
Hi guys, any update on this? Has anyone been able to reliably use VRAM as swap?

**Comment by bisqwit** [comment]
It only works if the following two conditions are met:
1) The GPU driver code/data is never put in swap
2) The vramfs driver code/data is never put in swap.
If you somehow can guarantee these aspects, then using VRAM as swap will work.

**Comment by montvid** [comment]
Did not work for me the one time I tried it. Seems the project is abandoned...

**Comment by wonghang** [comment]
fuse should be able not to swap itself. But I attempted to add `mlockall()` in vramfs code, it didn't work either. It appears that GPU driver (nvidia) was swapped.

In nvidia driver, there are some undocumented functions (prefix by `rm_`, run `cat /proc/kallsyms | grep nvidia` to see) to access GPU memory. I think they are parts of GPUDirect RDMA (https://docs.nvidia.com/cuda/gpudirect-rdma/index.html).
If we can somehow hack them and write a kernel driver to handle the paging, it may be possible to use GPU as swap.

**Comment by Atrate** [comment]
It is possible to achieve this, see https://wiki.archlinux.org/title/Swap_on_video_RAM , section FUSE.

> The vramfs driver code/data is never put in swap.
This can be achieved with https://wiki.archlinux.org/title/Swap_on_video_RAM#Complete_system_freeze_under_high_memory_pressure

I tested it under high memory pressure (`stress -m 10 --vm-bytes 3G --vm-hang 10` on a 32G system) and it didn't fall over, but only after applying the aforementioned fix.

**Comment by bisqwit** [comment]
This looks like a proper solution indeed.

**Comment by Atrate** [comment]
I've tried implementing `mlockall`. If you want to, you can test whether it works for you and fixes deadlocks without needing to use a systemd service.

https://github.com/Overv/vramfs/pull/32

**Comment by twobombs** [comment]
I would like to add to this discussion that the addition of vramfs as a block device would help using vramfs as a dedicated L2ARC ZFS buffer. 

We are using very big dedicated nvme swap raid arrays for quantum computing and need something that is faster then 8-16 NVME sticks in RAID to collect the IO in a buffer that is not in main memory. 

We make use of a lot of (virtual) memory so an L2ARC buffer in vram would be awesome; the GPUs would get a new lease on life because we went to CPU only calculation because of huge memory requirements to store the eigen vector (think 8/16TB)

**Comment by Atrate** [comment]
> I would like to add to this discussion that the addition of vramfs as a block device would help using vramfs as a dedicated L2ARC ZFS buffer.
> 
> We are using very big dedicated nvme swap raid arrays for quantum computing and need something that is faster then 8-16 NVME sticks in RAID to collect the IO in a buffer that is not in main memory.
> 
> We make use of a lot of (virtual) memory so an L2ARC buffer in vram would be awesome; the GPUs would get a new lease on life because we went to CPU only calculation because of huge memory requirements to store the eigen vector (think 8/16TB)

@twobombs 

You can make a loop device with `losetup` but NVME RAID will probably be faster than vramswap, the performance is still somewhat lacking in certain areas.

**Comment by twobombs** [comment]
@Atrate thank you very much for the loop solution. will look into this and if ZFS will allow a loop device as cache. the swap I/O usage pattern is random read/write, not stream. a PCIe VRAM device might offer better speeds whilst at the same time making the workload on NVME raid devices more 'stream'-lined when changes are comitted to the array.

**Comment by 5l1v3r1** [comment]
> It is possible to achieve this, see https://wiki.archlinux.org/title/Swap_on_video_RAM , section FUSE.
> 
> > The vramfs driver code/data is never put in swap.
> 
> This can be achieved with https://wiki.archlinux.org/title/Swap_on_video_RAM#Complete_system_freeze_under_high_memory_pressure
> 
> I tested it under high memory pressure (`stress -m 10 --vm-bytes 3G --vm-hang 10` on a 32G system) and it didn't fall over, but only after applying the aforementioned fix.

This is genius.

**Comment by twobombs** [comment]
I went a step further and added VRAM cache files for ZFS based SWAP.
It is fairly hilarious to see IO come through on NVTOP

![Screenshot_from_2023-02-01_20-10-18](https://user-images.githubusercontent.com/12692227/216598936-1522724f-e5ea-4c0d-a1ff-2f30d90373b6.png)

**Comment by aedalzotto** [comment]
> It is possible to achieve this, see https://wiki.archlinux.org/title/Swap_on_video_RAM , section FUSE.
> 
> > The vramfs driver code/data is never put in swap.
> 
> This can be achieved with https://wiki.archlinux.org/title/Swap_on_video_RAM#Complete_system_freeze_under_high_memory_pressure
> 
> I tested it under high memory pressure (`stress -m 10 --vm-bytes 3G --vm-hang 10` on a 32G system) and it didn't fall over, but only after applying the aforementioned fix.

The solution seems to work for me, but when I increase swappiness from 10 to 180, it simply freezes.
The same happens without increasing swappiness when running `mprime`.

I am running vramfs as a service, as the workaround cited above suggests. The only thing I think I am doing different is using a loopback, as my swapfile is being created with holes.

Does anyone have an idea of what is happening?

**Comment by fanzhuyifan** [comment]
> The solution seems to work for me, but when I increase swappiness from 10 to 180, it simply freezes. The same happens without increasing swappiness when running `mprime`.
> 
> I am running vramfs as a service, as the workaround cited above suggests. The only thing I think I am doing different is using a loopback, as my swapfile is being created with holes.
> 
> Does anyone have an idea of what is happening?

What is the size of the vramfs you created and what is the size of the swap file you created? The latter needs to be bigger than the former, you need some extra room for the overhead of the filesystem (I think). In my setup, I used `2000MB` for the size of the vramfs, and `1600MB` for the size of the swapfile. I observed that my system freezes up when the size of the vramfs is too small.

**Comment by Atrate** [comment]
In reply to: https://github.com/Overv/vramfs/issues/3#issuecomment-1663130681

As suggested by [Fanzhuyifan](https://wiki.archlinux.org/title/User:Fanzhuyifan) I think that may be due to *other* GPU-management processes/libraries getting swapped out and maaaybe a fix is possible with a lot of systemd unit editing but that'd require tracking down every single library and process that is required for the operation of a dGPU and that seems like a chore.

**Comment by fanzhuyifan** [comment]
According to the documentation of [mlockall](https://man7.org/linux/man-pages/man2/mlock.2.html), 

> mlockall() locks all pages mapped into the address space of the
       calling process.  This includes the pages of the code, data, and
       stack segment, as well as shared libraries, user space kernel
       data, shared memory, and memory-mapped files.  All mapped pages
       are guaranteed to be resident in RAM when the call returns
       successfully; the pages are guaranteed to stay in RAM until later
       unlocked.
 
So shared libraries being swapped out should not be the reason of system freezes.

**Comment by Atrate** [comment]
> Edit: Examining the resident size and virtual memory size of the vramfs process, I think the issue is that vramfs asks for additional memory to serve reads/writes.

Is it? `mlockall` is called with the `MCL_CURRENT | MCL_FUTURE` flags, so it should also prevent all future allocations of memory from being swapped, unless I misunderstood the documentation.

Code in vramfs: https://github.com/Overv/vramfs/blob/829b1f2c259da2eb63ed3d4ddef0eeddb08b99e4/src/vramfs.cpp#L534

Documentation:

```
    MCL_CURRENT
              Lock all pages which are currently mapped into the address
              space of the process.

       MCL_FUTURE
              Lock all pages which will become mapped into the address
              space of the process in the future.  These could be, for
              instance, new pages required by a growing heap and stack
              as well as new memory-mapped files or shared memory
              regions.

```

**Comment by fanzhuyifan** [comment]
> > Edit: Examining the resident size and virtual memory size of the vramfs process, I think the issue is that vramfs asks for additional memory to serve reads/writes.
> 
> Is it? `mlockall` is called with the `MCL_CURRENT | MCL_FUTURE` flags, so it should also prevent all future allocations of memory from being swapped, unless I misunderstood the documentation.

Here are the steps to prove my point (on linux):

1. Start vramfs, say creating a filesystem with size 2000MB, and find the PID of the process.
2. Run `cat /proc/PID/status | grep Vm` to find the memory information. On a particular run on my computer I got

> VmPeak:	 7060808 kB
VmSize:	 7060808 kB
VmLck:	 6990308 kB
VmPin:	       0 kB
VmHWM:	  275588 kB
VmRSS:	  275588 kB
VmData:	  144976 kB
VmStk:	     164 kB
VmExe:	     132 kB
VmLib:	   14156 kB
VmPTE:	     628 kB
VmSwap:	       0 kB


3. Write random data to a file on the vramfs, and check memory usage again.
First run `dd if=/dev/random of=/tmp/vram/swapfile bs=1M count=1000`, and then I got:

> **VmPeak**:	 7585096 kB
**VmSize**:	 7388488 kB
**VmLck**:	 7317988 kB
VmPin:	       0 kB
**VmHWM**:	  286148 kB
**VmRSS**:	  286148 kB
**VmData**:	  156092 kB
VmStk:	     164 kB
VmExe:	     132 kB
VmLib:	   14156 kB
VmPTE:	     668 kB
VmSwap:	       0 kB

Note that the bolded entries all increased.

4. Let's read that file and check memory usage again. 
First run `sha256sum /tmp/vram/swapfile`, and then I got 

> VmPeak:	 7585096 kB
**VmSize**:	 7462220 kB
**VmLck**:	 7391720 kB
VmPin:	       0 kB
**VmHWM**:	  296072 kB
**VmRSS**:	  296072 kB
**VmData**:	  165960 kB
VmStk:	     164 kB
VmExe:	     132 kB
VmLib:	   14156 kB
VmPTE:	     692 kB
VmSwap:	       0 kB

The bolded entries increased again.

I believe this proves that vramfs asks for more memory when serving read and write requests.
I suspect that this is the reason the computer freezes when using vramfs as swap, even with the `mlockall` call.

**Comment by jnturton** [comment]
> Since there is already no available memory, the system freezes.

Wouldn't we see OOM Killer enties in the kernel logs in this case?

**Comment by mykaul** [comment]
In my tests, before trying to get it to be used as swap, I've tested 4K write performance - I could not get reasonable performance:
- Host NVMe (no FUSE): 1431 MiB/s, 366k IOPS (4K DIO writes)
- Current vramfs DIO: 52.8 MiB/s, 13.5k IOPS
- Current vramfs buffered: 120 MiB/s, 30.6k IOPS
